qbraid_qir/qasm3/visitor.py

```diff
@@ -133,17 +145,38 @@ def entry_point(self) -> str:
     def finalize(self) -> None:
         self._builder.ret(None)
 
-    def _set_scope(self, scope: Scope) -> None:
-        self._scope = scope
+    def _push_scope(self, scope: dict) -> None:
+        self._scope.append(scope)
+
+    def _pop_scope(self) -> None:
+        if len(self._scope) > 0:
+            self._scope.pop()
 
-    def _in_gate(self) -> bool:
-        return self._scope == Scope.GATE
+        raise IndexError("Scope list is empty, can not pop")
```

Comment:
This implementation will raise an `IndexError` regardless of whether the `_scope` list is empty or not. Should instead be something like

```python
def _pop_scope(self) -> None:
    if len(self._scope) == 0:
        raise IndexError("Scope list is empty, cannot pop")
    self._scope.pop()
```

Important to have a test case that hits this.